### PR TITLE
Graphics Buffer Update -- New PR

### PR DESCRIPTION
**Pull request by rburema**
_The HD of my private laptop is fried. While I resolve that, it's easier to work on the laptop from work. Since switching accounts back and forth can be a big pain and lead to confusion, I'm developing these from my work-account._

_I now return you the text of the original PR_:

The highly interdependent Issues this fixes:

The Vertex Array Object (VAO) for a render wasn't actually being (re)used, but regenerated each time. Now it's cached.
While (otherwise) vertex- and index-buffers where already cached for normal rendering, the index-buffer for a mesh was recreated each time only a part of the mesh (a range) needed to be drawn (for example, Layer and Simulation view in Cura). (NB: This was due to a workaround on our side, caused by an only partial implementation of a specific bit of the OpenGL API by the Python interface we're using.) This is now fixed.
Complications:

In order to show 'ranged'/partial meshes, implementing programs (like Cura) must now have shaders that take a draw-range uniform, which can be compared against the newly introduced 'vertex-index' attribute. (As we can't use the OpenGL API due to partial implementation, and we can't use the gl_primitiveID built-in due to support of legacy shaders.) See also Graphic Buffers Update ~original: Ultimaker/Cura#10188~ new: Ultimaker/Cura#10259, the front-end implementation of this PR.
See also this related discussion #664 earlier.

### CODE REVIEW

**Comment by rburema** [comment]
So I finally properly ran the profile that @nallath suggested.

For a small case the answer is straightforward:

Current main branch is a bit slower:
old_code_small_example

Code in PR (I've merged main into it, will commit soon), is a bit faster:
new_code_small_example

For a larger example, I noticed something a bit strange; the profile shows a _huge_ win for this PR, but ... the only difference I noticed as a user is that when in preview mode (where the speedup should be most noticeable, if at all), and I only selected one, or a few, layers, the 'new way' was still chugging a bit, when the 'old way' ran smoothly. When the entire model was shown on screen there wasn't much difference, I _think_.

This can partly be explained in that the new way of doing things is slightly more reliant on the graphics card rather than the normal compute cores, and that the whole model is loaded at all times, just not all parts are _shown_ (early out in shader when a part shouldn't be shown of course). But it needs to be able to render the entire model _anyway_, so I don't consider this a huge loss.

Anyway, here is the chart for the current main:
old_code_big_example

... and for the apparently +-10x as fast new code:
new_code_big_example

File: UM/Mesh/MeshData.py [hunk 1]
```diff
@@ -349,6 +350,12 @@ def getAttribute(self, key: str):
 
         return self._attributes[key]
 
+    def getCachedUserValue(self, key: str) -> Any:
+        return None if key not in self._user_data_cache else self._user_data_cache[key]
```
Comment (nallath) [inline UM/Mesh/MeshData.py:354]:
```suggestion
        return self._user_data_cache.get(key) 
```

File: UM/Mesh/MeshData.py [hunk 2]
```diff
@@ -349,6 +350,12 @@ def getAttribute(self, key: str):
...
+
+    def setCachedUserValue(self, key: str, value: Any):
```
Comment (nallath) [inline UM/Mesh/MeshData.py:356]:
```suggestion
    def setCachedUserValue(self, key: str, value: Any) -> None:  
```

**Review by nallath — COMMENTED** [review]
I've got some tiny nitpicks, but it looks good otherwise.

**Review by Ghostkeeper — DISMISSED** [review]

File: UM/View/RenderBatch.py [hunk 3]
```diff
@@ -302,21 +261,79 @@ def _renderItem(self, item: Dict[str, Any]):
...
+    def _vertexBuffersSetup(self, mesh: MeshData) -> Optional[QOpenGLVertexArrayObject]:
+        # See if the mesh has already been stored to the GPU:
+        vao = cast(Optional[QOpenGLVertexArrayObject], mesh.getCachedUserValue(self._shader.getReferenceKey()))
+        if vao is not None:
+            return vao
 
-        vertex_buffer.release()
+        # Initialize VAO (VertexArrayObject). On activation, this will wrap around the other vertex/index buffers.
+        # That enables reusing them without much fuss.
+        if OpenGLContext.properties["supportsVertexArrayObjects"]:
```
[the remark this answers is not in the document]
Comment (rburema) [reply UM/View/RenderBatch.py:272]: Good point, it should return None if it _isn't_ supported, then continue. After all, VAO stands for Vertex Array Object.

File: UM/View/RenderBatch.py [hunk 4]
```diff
@@ -302,21 +261,79 @@ def _renderItem(self, item: Dict[str, Any]):
...
+        mesh = cast(MeshData, item["mesh"])
+        if mesh.getVertexCount() == 0:
+            return
+
+        if self._render_range is not None:
+            self._shader.setUniformValue("u_drawRange", [self._render_range[0], self._render_range[1]])
+            self._shader.setUniformValue("draw_range", [self._render_range[0], self._render_range[1]])
+        else:
+            self._shader.setUniformValue("u_drawRange", [-1.0, -1.0])
+            self._shader.setUniformValue("draw_range", [-1.0, -1.0])
```
[the remark this answers is not in the document]
Comment (rburema) [reply UM/View/RenderBatch.py:306]: Well, the draw-range is sometimes not None, so they already do! :-)

File: UM/View/RenderBatch.py [hunk 5]
```diff
@@ -302,21 +261,79 @@ def _renderItem(self, item: Dict[str, Any]):
...
+        return vao
+
+    def _renderItem(self, item: Dict[str, Any]) -> None:
+        mesh = cast(MeshData, item["mesh"])
+        if mesh.getVertexCount() == 0:
+            return
+
+        if self._render_range is not None:
+            self._shader.setUniformValue("u_drawRange", [self._render_range[0], self._render_range[1]])
+            self._shader.setUniformValue("draw_range", [self._render_range[0], self._render_range[1]])
```
[the remark this answers is not in the document]
Comment (rburema) [reply UM/View/RenderBatch.py:303]: One of them is enough. I must've forgotten to remove the other one.

**Review by Ghostkeeper — APPROVED** [review]

**Comment by Ghostkeeper** [comment]
This PR is on hold for the moment since the companion PR in the Cura repository has encountered a bug while testing.

**Pull request merged**